core/trino-main/src/main/java/io/trino/server/TaskResource.java

```diff
@@ -516,4 +496,31 @@ private static Duration randomizeWaitTime(Duration waitTime)
         long halfWaitMillis = waitTime.toMillis() / 2;
         return new Duration(halfWaitMillis + ThreadLocalRandom.current().nextLong(halfWaitMillis), MILLISECONDS);
     }
+
+    private static Response createBufferResultResponse(SqlTaskManager.SqlTaskWithResults taskWithResults, BufferResult result)
+    {
+        // This response may have been created as the result of a timeout, so refresh the task heartbeat
+        taskWithResults.recordHeartbeat();
+
+        List<Slice> serializedPages = result.getSerializedPages();
+
+        GenericEntity<?> entity = null;
+        Status status;
+        if (serializedPages.isEmpty()) {
+            status = Status.NO_CONTENT;
+        }
+        else {
+            entity = new GenericEntity<>(serializedPages, new TypeToken<List<Slice>>() {}.getType());
+            status = Status.OK;
+        }
+
+        return Response.status(status)
+                .entity(entity)
+                .header(TRINO_TASK_INSTANCE_ID, result.getTaskInstanceId())
+                .header(TRINO_PAGE_TOKEN, result.getToken())
+                .header(TRINO_PAGE_NEXT_TOKEN, result.getNextToken())
+                .header(TRINO_BUFFER_COMPLETE, result.isBufferComplete())
+                .header(TRINO_TASK_FAILED, taskWithResults.isTaskFailed())
```

Comment:
nit: maybe add a comment here, something along the line "check task failure after getting the result"